apispec/ext/marshmallow/swagger.py

```diff
@@ -348,9 +350,10 @@ def fields2parameters(fields, schema=None, spec=None, use_refs=True, dump=True,
             default_in=default_in
         )
             for field_name, field_obj in iteritems(fields)
-            if (
-                (field_name not in exclude_fields)
-                and not (field_obj.dump_only and not dump)
+            if not (
+                field_name in exclude_fields or
+                (field_name in dump_only_fields or field_obj.dump_only) and not dump or 
+                (field_name in load_only_fields or field_obj.load_only) and dump
```

Comment:
I would change the order of the conditions and add more parentheses (just to avoid possible typo-bugs in future):

```python
if not (
    field_name in exclude_fields or
    (not dump and (field_obj.dump_only or field_name in dump_only_fields)) or
    (dump and (field_obj.load_only or field_name in load_only_fields))
)
```

Lazy evaluation of conditions will help to avoid unnecessary `in` checking.

Other than that, I think it is a good change. Well done!